File: ObeliskGemEV/event/stats.py
"""
Player and Enemy statistics for Event Simulator.
"""

from dataclasses import dataclass


@dataclass
class PlayerStats:
    """Player statistics for event simulation"""
    default_atk_time: float = 2.0
    default_walk_time: float = 4.0  # Time to walk between enemies (matches Lua original: defaultWalkTime=4)
    walk_speed: float = 1.0
    atk_speed: float = 1.0
    health: int = 100
    atk: int = 10
    crit: int = 0
    crit_dmg: float = 2.0
    block_chance: float = 0.0
    game_speed: float = 1.0
    prestige_bonus_scale: float = 0.1
    x2_money: int = 0
    x5_money: int = 0
    max_wave: int = 1


@dataclass
class EnemyStats:
    """Enemy statistics for event simulation
    
    Scaling per wave:
    - HP = base_health + health_scaling * wave
    - ATK = max(1, atk + atk_scaling * wave)
    - ATK Speed = atk_speed + 0.02 * wave
    - Crit Chance = crit + wave
    - Crit Dmg = crit_dmg + crit_dmg_scaling * wave
    """
    default_atk_time: float = 2.0
    atk_speed: float = 0.8
    base_health: int = 4
    health_scaling: int = 7
    atk: float = 2.5
    atk_scaling: float = 0.6
    crit: int = 0
    crit_dmg: float = 1.0
    crit_dmg_scaling: float = 0.05

File: web/src/lib/event/stats.ts
// Ported from ObeliskGemEV/event/stats.py

export interface PlayerStats {
  defaultAtkTime: number; // seconds
  defaultWalkTime: number; // seconds
  walkSpeed: number;
  atkSpeed: number; // attacks per second
  health: number;
  atk: number;
  crit: number; // percent
  critDmg: number; // multiplier
  blockChance: number; // 0..1
  gameSpeed: number; // multiplier-ish
  prestigeBonusScale: number;
  x2Money: number;
  x5Money: number; // percent
}

export interface EnemyStats {
  defaultAtkTime: number;
  atkSpeed: number;
  baseHealth: number;
  healthScaling: number;
  atk: number;
  atkScaling: number;
  crit: number; // percent base
  critDmg: number;
  critDmgScaling: number;
}

export function createBasePlayerStats(): PlayerStats {
  return {
    defaultAtkTime: 2.0,
    defaultWalkTime: 4.0,
    walkSpeed: 1.0,
    atkSpeed: 1.0,
    health: 100,
    atk: 10,
    crit: 0,
    critDmg: 2.0,
    blockChance: 0.0,
    gameSpeed: 1.0,
    prestigeBonusScale: 0.1,
    x2Money: 0,
    x5Money: 0,
  };
}

export function createBaseEnemyStats(): EnemyStats {
  return {
    defaultAtkTime: 2.0,
    atkSpeed: 0.8,
    baseHealth: 4,
    healthScaling: 7,
    atk: 2.5,
    atkScaling: 0.6,
    crit: 0,
    critDmg: 1.0,
    critDmgScaling: 0.05,
  };
}

